hbase-testing-util/src/test/java/org/apache/hadoop/hbase/testing/TestTestingHBaseClusterImplForCPs.java

```diff
@@ -0,0 +1,107 @@
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertSame;
+
+import java.io.IOException;
+import java.util.List;
+import org.apache.hadoop.hbase.HBaseClassTestRule;
+import org.apache.hadoop.hbase.ServerName;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.client.Admin;
+import org.apache.hadoop.hbase.client.ColumnFamilyDescriptorBuilder;
+import org.apache.hadoop.hbase.client.Connection;
+import org.apache.hadoop.hbase.client.ConnectionFactory;
+import org.apache.hadoop.hbase.client.RegionInfo;
+import org.apache.hadoop.hbase.client.RegionInfoBuilder;
+import org.apache.hadoop.hbase.client.RegionLocator;
+import org.apache.hadoop.hbase.client.TableDescriptorBuilder;
+import org.apache.hadoop.hbase.regionserver.OnlineRegions;
+import org.apache.hadoop.hbase.regionserver.Region;
+import org.apache.hadoop.hbase.testclassification.LargeTests;
+import org.apache.hadoop.hbase.testclassification.MiscTests;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import org.apache.hbase.thirdparty.com.google.common.io.Closeables;
+
+@Category({ MiscTests.class, LargeTests.class })
+public class TestTestingHBaseClusterImplForCPs {
+
+  @ClassRule
+  public static final HBaseClassTestRule CLASS_RULE =
+    HBaseClassTestRule.forClass(TestTestingHBaseClusterImplForCPs.class);
+
+  private static TestingHBaseCluster CLUSTER;
+
+  private static TableName NAME = TableName.valueOf("test");
+
+  private static byte[] CF = Bytes.toBytes("cf");
+
+  private static Connection CONN;
+
+  private static Admin ADMIN;
+
+  @BeforeClass
+  public static void setUpBeforeClass() throws Exception {
+    CLUSTER = TestingHBaseCluster.create(TestingHBaseClusterOption.builder().numMasters(2)
+      .numRegionServers(3).numDataNodes(3).build());
+    CLUSTER.start();
+    CONN = ConnectionFactory.createConnection(CLUSTER.getConf());
+    ADMIN = CONN.getAdmin();
+    ADMIN.createTable(TableDescriptorBuilder.newBuilder(NAME)
+      .setColumnFamily(ColumnFamilyDescriptorBuilder.of(CF)).build());
+    ADMIN.balancerSwitch(false, true);
+  }
+
+  @AfterClass
+  public static void tearDownAfterClass() throws Exception {
+    Closeables.close(ADMIN, true);
+    Closeables.close(CONN, true);
+    if (CLUSTER.isClusterRunning()) {
+      CLUSTER.stop();
+    }
+  }
+
+  @Test
+  public void testGetRegion() throws IOException {
+    List<RegionInfo> infos = ADMIN.getRegions(NAME);
+    assertEquals(1, infos.size());
+    RegionInfo info = infos.get(0);
+    Region region = CLUSTER.getRegion(info).get();
+    ServerName loc;
+    try (RegionLocator locator = CONN.getRegionLocator(NAME)) {
+      loc = locator.getRegionLocation(info.getStartKey()).getServerName();
+    }
+    OnlineRegions onlineRegionsInterface = CLUSTER.getOnlineRegionsInterface(loc).get();
```

Comment:
I'm confused why this isn't a ClassCastException going between HRegionServer and OnlineRegions but I see the test is passing in at least some of the test runs. Is there something interesting going on that I'm not understanding?

Reply:
As said above, the HRegionServer just extends the OnlineRegions interface, so there is no problem. So actually in the CP implementation, when you get an OnlineRegions instance, it is actually a HRegionServer :)